Finish first version of tool

blazejulanowicz opened the pull request:
We've completed first version of raw image data previewer.
Finished tasks:
- Implement the previewer for the mostly used RGB formats, make it abstract so adding new RGB formats is simple
- Add an option to preview common YUV formats
- Add an CLI util to automatically convert the raw data to a higher-level format (e.g. png)
- Add planar YUV formats support

One task was abandoned - _Add option to define custom RGB/YUV formats_.
Instead we have implemented long list of most popular/most supported formats.

Documentation can be found in README and inside code files. Full list of formats can be found under `docs/SUPPORTED_FORMATS.md`.

tgorochowik commented:
Thanks! Looks like a ton of great work! I will try to find time to play with it a little bit more before adding a proper review.

I have one question for now:
> One task was abandoned - Add option to define custom RGB/YUV formats.
Instead we have implemented long list of most popular/most supported formats.

Will this be easy to add if it is needed in the future? We asked for this as every once in a while there is a camera that uses a really weird and non standard format. There are also sometimes situations where the data is corrupted, being able to play with the format could be useful for debugging in those cases.

blazejulanowicz commented:
> > One task was abandoned - Add option to define custom RGB/YUV formats.
> > Instead we have implemented long list of most popular/most supported formats.
> 
> Will this be easy to add if it is needed in the future? We asked for this as every once in a while there is a camera that uses a really weird and non standard format. There are also sometimes situations where the data is corrupted, being able to play with the format could be useful for debugging in those cases.

More or less it already works for less standard types. For example, parsing any RGB-like format should be currently possible - even 1 bit per pixel component. Of course, support is only provided for listed color components formats (RGBA/BGRA and ARGB/ABGR), but even that can be quickly adjusted by making proper binding to _opencv_ conversion type. Besides this, you should be able to create any color format you'd like (but be aware, that system testing was done only for the listed format).

When it comes to BayerRGB, it supports any length of color component, provided every color components has n-bytes for itself (by that I mean that i. e. 12bit color component is fitted into 2 bytes cell and empty bits are zeros). Fully packed color formats like `V4L2_PIX_FMT_SRGGB10P` are not supported, even with addition of `_parse_not_bytefilled` function. They just need completely different approach to parse those bytes.

Grayscale and YUV color formats can be expanded by `_parse_not_bytefilled` function, which is made to extract smaller color components from byte or a few consecutive bytes (i. e. RGB332 or RGB565 - their components do not have "byte only for themselves"). It SHOULD work out-of-the-box for grayscale images. For YUV images there should be also added color components values normalization, because for now only 8-bit YUVs are supported, so this wasn't needed.

Other than that basically only new pixel formats can be added - but it's painless to implement if there is already conversion from this kind of color component alignment implemented in OpenCV (i. e. COLOR_BAYER_BG2RGB). It's just adding new switch case in `get_displayable` function.

The only thing that can be pretty hard to parametrize is subsampling in YUV formats, but I think it's doable by modifying code (applying parameters instead of hardcoded numbers) of `ParserYUV422Planar`.

tgorochowik commented on app/__main__.py:
```diff
@@ -0,0 +1,52 @@
+    img = load_image(args["FILE_PATH"], args["color_format"],
+                     args["resolution"])
+    save_image_as_file(get_displayable(img), args["export"])
+"""
+cv.imshow(args["FILE_PATH"], get_displayable(img))
+cv.waitKey(0)
+cv.destroyAllWindows()
+"""
```
Should this be removed?

blazejulanowicz replied:
removed

tgorochowik commented on docs/SUPPORTED_FORMATS.md:
```diff
@@ -0,0 +1,51 @@
+### RGB-like
+
+| Name    |     VL42 Identifier    |   current support  |   future support   |
```
is the `future support` column still needed?

blazejulanowicz replied:
column removed

tgorochowik reviewed (commented):
Looks good in general! Great work guys!

There are still some situations where it is pretty slow and not very stable.

For example:
* open the app
* open the `ABGR32_1000_750` file from tests
* set the width to `1000`
* set the format to `ABGR32` (everything works fine so far)
* change the format to `RGBA444` the application seems to get completely stuck - it should display a malformed image

Could you please take a look at that? And if possible - add some tests to make sure the app doesn't crash.

btw what exactly is RGBA444 - if it has 4 color components shouldn't it be rgba4444? or if it only has 3 compontents it should be rgb444

blazejulanowicz commented:
> There are still some situations where it is pretty slow and not very stable.
> 
> For example:
> 
>     * open the app
> 
>     * open the `ABGR32_1000_750` file from tests
> 
>     * set the width to `1000`
> 
>     * set the format to `ABGR32` (everything works fine so far)
> 
>     * change the format to `RGBA444` the application seems to get completely stuck - it should display a malformed image
> 
> 
> Could you please take a look at that? And if possible - add some tests to make sure the app doesn't crash.

We fixed that but forgot to finally add. It's because of PIL way of transforming image with alpha channel and it's finally not very efficient - it's not a problem of a backend.
Of course for formats that has to be analyzed more deeply (because color components do not have "their own byte") parsing wil last a little bit longer than for normal, integer ones.

> 
> btw what exactly is RGBA444 - if it has 4 color components shouldn't it be rgba4444? or if it only has 3 compontents it should be rgb444

I'm not the one naming the color formats: https://linuxtv.org/downloads/v4l-dvb-apis/userspace-api/v4l/pixfmt-rgb.html#

tgorochowik reviewed (approved):
Fair enough, let's keep the naming scheme consistent with what you based them on (we may want to reconsider the naming scheme later but it's okay for now).

The performance can probably be addressed separately too, let's merge this initial release!

Thanks for all the great work!
